src/Orleans.Runtime/GrainDirectory/GrainDirectoryPartition.cs

```diff
@@ -264,7 +262,8 @@ internal int GetGrainETag(GrainId grain)
             {
                 foreach (var pair in other.partitionData)
                 {
-                    if (partitionData.TryGetValue(pair.Key, out var existing))
+                    ref var grainInfo = ref CollectionsMarshal.GetValueRefOrAddDefault(partitionData, pair.Key, out _);
+                    if (grainInfo is { } existing)
```

Comment:
This works given `GrainInfo` is a class, but checking the currently ignored `exists` out-param from `GetValueRefOrAddDefault` would be more robust to refactoring